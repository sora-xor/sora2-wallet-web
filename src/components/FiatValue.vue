<template>
  <span class="fiat-value" v-if="value">
    <span class="fiat-value__prefix">~$</span>
    <span class="fiat-value__number">{{ formatted.value }}</span>
    <span class="fiat-value__decimals">{{ formatted.decimals }}</span>
  </span>
</template>

<script lang="ts">
import { Vue, Component, Prop } from 'vue-property-decorator'

import { FPNumber } from '@sora-substrate/util'

@Component
export default class FiatValue extends Vue {
  @Prop({ default: '', type: String }) readonly value!: string
  @Prop({ default: true, type: Boolean }) readonly withDecimals!: boolean

  get formatted (): any {
    const delimiterIndex = this.value.indexOf(FPNumber.DELIMITERS_CONFIG.decimal)
    if (this.withDecimals) {
      return {
        value: delimiterIndex !== -1 ? this.value.substring(0, delimiterIndex) : this.value,
        decimals: delimiterIndex !== -1 ? this.value.substring(delimiterIndex) : FPNumber.DELIMITERS_CONFIG.decimal + '00'
      }
    }
    return {
      value: delimiterIndex !== -1 ? this.value.substring(0, delimiterIndex) : this.value,
      decimals: ''
    }
  }
}
</script>

<style scoped lang="scss">
.fiat-value {
  color: var(--s-color-fiat-value);
  font-family: var(--s-font-family-default);
  font-weight: 800;
  font-size: var(--s-font-size-small);
  line-height: var(--s-line-height-medium);
  letter-spacing: var(--s-letter-spacing-small);
  &__prefix {
    opacity: .6;
    padding-right: calc(var(--s-basic-spacing) / 4);
  }
  &__number {
    font-weight: 600;
  }
  &__decimals {
    font-weight: 300;
    font-size: var(--s-font-size-mini);
  }
}
</style>
