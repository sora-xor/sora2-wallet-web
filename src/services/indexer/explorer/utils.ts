import { formatStringNumber } from '../../../util';

import type { SubqueryAssetEntity } from '../subquery/types';
import type { SubsquidAssetEntity } from '../subsquid/types';
import type { FiatPriceObject, UpdatesStream, PoolApyObject } from '../types';

export function parseAssetFiatPrice(entity: SubsquidAssetEntity | SubqueryAssetEntity): FiatPriceObject {
  const acc = {};
  const id = entity.id;
  const priceFPNumber = formatStringNumber(entity.priceUSD);
  const isPriceFinity = priceFPNumber.isFinity();
  if (isPriceFinity) {
    acc[id] = priceFPNumber.toCodecString();
  }
  return acc;
}

export function parsePriceStreamUpdate(entity: UpdatesStream): Nullable<FiatPriceObject> {
  if (!entity?.data) return null;

  const data = JSON.parse(entity.data);

  return Object.entries(data).reduce((acc, [id, price]) => {
    const priceFPNumber = formatStringNumber(price as string);
    const isPriceFinity = priceFPNumber.isFinity();
    if (isPriceFinity) {
      acc[id] = priceFPNumber.toCodecString();
    }
    return acc;
  }, {});
}

export function parseApyStreamUpdate(entity: UpdatesStream): PoolApyObject {
  const data = entity?.data ? JSON.parse(entity.data) : {};

  return Object.entries(data).reduce((acc, [id, apy]) => {
    const strategicBonusApyFPNumber = formatStringNumber(apy as string);
    const isStrategicBonusApyFinity = strategicBonusApyFPNumber.isFinity();
    if (isStrategicBonusApyFinity) {
      acc[id] = strategicBonusApyFPNumber.toCodecString();
    }
    return acc;
  }, {});
}
