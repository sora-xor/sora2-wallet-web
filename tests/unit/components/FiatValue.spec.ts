import { shallowMount } from '@vue/test-utils'

import { useDescribe, localVue } from '../../utils'
import FiatValue from '@/components/FiatValue.vue'

useDescribe('FiatValue.vue', FiatValue, () => {
  it('should be rendered correctly', () => {
    const propsData = {
      value: '1234.5678',
      withDecimals: true
    }
    const wrapper = shallowMount(FiatValue, { localVue, propsData })
    expect(wrapper.element).toMatchSnapshot()
  })
})
