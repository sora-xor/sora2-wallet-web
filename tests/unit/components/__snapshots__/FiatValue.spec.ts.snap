// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FiatValue.vue should be rendered correctly 1`] = `
<span
  class="fiat-value"
>
  <span
    class="fiat-value__prefix"
  >
    ~$
  </span>
   
  <span
    class="fiat-value__number"
  >
    1,234
  </span>
   
  <span
    class="fiat-value__decimals"
  >
    .56
  </span>
</span>
`;
